### PR TITLE
feat: add implementation of `ToRedisArgs`,`FromRedisValue` traits for `Arc<T>`,`Box<T>`,`Rc<T>`

### PR DESCRIPTION
**Pull request by xoac**
With these macros(kudos to serde crate) it will be very easy to add other std types like `Mutex`,`Rc`,`ArcWeak` 

I think this PR might contains breaking changes for some users because from now you need to specify output type. See `test_vec` function.

### CODE REVIEW

File: redis/src/types.rs [hunk 1]
```diff
@@ -1460,17 +1460,50 @@ impl<T: ToRedisArgs> ToRedisArgs for Option<T> {
...
+deref_impl! {
+    <'a, T: ?Sized> ToRedisArgs for &'a mut T where T: ToRedisArgs
+}
+
+deref_impl! {
+    <T: ?Sized> ToRedisArgs for Box<T> where T: ToRedisArgs
+}
+
+deref_impl! {
+    /// Encoding a data structure containing `Arc` will encode a copy of
```
Comment (nihohit) [inline redis/src/types.rs:1502]: I don't understand this. encoding in this library always incurs a copy of the content, what are you clarifying?

File: redis/src/types.rs [hunk 2]
```diff
@@ -1999,7 +2080,6 @@ macro_rules! from_vec_from_redis_value {
 
 from_vec_from_redis_value!(<T> Vec<T>);
 from_vec_from_redis_value!(<T> std::sync::Arc<[T]>);
-from_vec_from_redis_value!(<T> Box<[T]>; Vec::into_boxed_slice);
```
Comment (nihohit) [inline redis/src/types.rs:2002]: why prefer your implementation?

File: redis/src/types.rs [hunk 3]
```diff
@@ -1925,6 +1958,54 @@ impl FromRedisValue for String {
...
+            }
+
+            fn from_owned_redis_value(v: Value) -> RedisResult<Self> {
+                Box::from_owned_redis_value(v).map(Into::into)
+            }
+        }
+    };
+}
+
+box_forwarded_impl!(T, std::sync::Arc<T>);
```
Comment (nihohit) [inline redis/src/types.rs:2006]: why can't this use `forwarded_impl` with `Arc::new`?

File: redis/tests/test_types.rs [hunk 4]
```diff
@@ -77,6 +81,10 @@ fn test_i32() {
 
         let bad_i: Result<i32, _> = parse_mode.parse_redis_value(Value::SimpleString("42x".into()));
```
Comment (nihohit) [inline redis/tests/test_types.rs:82]: minor - can you move `Value::SimpleString("42x".into())` to a variable, just to be clear that both calls use the same value?

File: redis/tests/test_types.rs [hunk 5]
```diff
@@ -93,6 +101,33 @@ fn test_u32() {
...
+    for parse_mode in [RedisParseMode::Owned, RedisParseMode::Ref] {
+        let simple_string_exp = "Simple string".to_string();
+        let v = parse_mode.parse_redis_value(Value::SimpleString(simple_string_exp.clone()));
+        assert_eq!(v, Ok(Box::new(simple_string_exp.clone())));
+    }
+}
+
+#[test]
+fn test_parse_arc() {
+    use redis::Value;
```
Comment (nihohit) [inline redis/tests/test_types.rs:117]: I think we can move these using statements out of the tests, to the top of the file.

File: redis/tests/test_types.rs [hunk 6]
```diff
@@ -162,26 +197,26 @@ fn test_arc_slice() {
             Value::BulkString("2".into()),
             Value::BulkString("3".into()),
         ]));
-        assert_eq!(v, Ok(Arc::from(vec![1i32, 2, 3])));
```
Comment (nihohit) [inline redis/tests/test_types.rs:165]: why is this change required?

**Review by nihohit — COMMENTED** [review]
very impressive work, thanks!

File: redis/src/types.rs (hunk 2 above)
Comment (xoac) [reply redis/src/types.rs:2002]: I wanted to remove `from_vec_from_redis_value!` macro and use just `forwarded_impl!((T), std::sync::Arc<[T]>, std::sync::Arc::new);` but I wasn't able to make it compile. If you have suggestion how to make it compile this can benefit in one macro less.

File: redis/tests/test_types.rs (hunk 6 above)
Comment (xoac) [reply redis/tests/test_types.rs:165]: because you can have now `Arc<T>` and `Arc<[T]>` and you must give Rust a hint about type.

File: redis/tests/test_types.rs (hunk 4 above)
Comment (xoac) [reply redis/tests/test_types.rs]: Both of these values doesn't need to use the same value in this tests but refactor it anyway

File: redis/src/types.rs (hunk 1 above)
Comment (nihohit) [reply redis/src/types.rs]: you can reuse the output, but this macro is for the inputs :)
`ToRedisArgs` - inputs
`FromRedisValue` - outputs

File: redis/src/types.rs (hunk 2 above)
Comment (nihohit) [reply redis/src/types.rs]: I can try, but your macro doesn't cover `Vec`, so I don't see how we can avoid having 2 macros here.

File: redis/tests/test_types.rs (hunk 6 above)
Comment (nihohit) [reply redis/tests/test_types.rs]: this seems to break more tests than just this. if you remove the explicit support for `Arc<[T]>`, would it still be derived by the macros?

File: redis/src/types.rs (hunk 2 above)
Comment (xoac) [reply redis/src/types.rs]: The `from_vec_from_redis_value!` will be used only for `Vec<T>` so no need to keep macro just direct impl for Vec.

File: redis/tests/test_types.rs (hunk 6 above)
Comment (xoac) [reply redis/tests/test_types.rs]: it will not be delivered because implementation for `Arc<[T]>` is to create `Vec<T>` first and then use [Arc::from](https://doc.rust-lang.org/std/sync/struct.Arc.html#impl-From%3CVec%3CT,+A%3E%3E-for-Arc%3C%5BT%5D,+A%3E) as far as I understand.

Comment (xoac) [reply redis/tests/test_types.rs]: But I accidentally changed testing `Arc<[T]>` into `Arc<Vec<T>>` I fixed this with [`4f1d21e` (#1088)](https://github.com/redis-rs/redis-rs/pull/1088/commits/4f1d21ea4c086e368e7420063e2d5ad9e266fce6) commit

File: redis/src/types.rs (hunk 2 above)
Comment (xoac) [reply redis/src/types.rs:2002]: Was able to remove from_vec_from_redis_value! macro with [`f91f8e0` (#1088)](https://github.com/redis-rs/redis-rs/pull/1088/commits/f91f8e0e7fefbd4e070a9fccf48f16c385d9c2bd) commit

File: redis/src/types.rs [hunk 7]
```diff
@@ -1925,81 +1954,92 @@ impl FromRedisValue for String {
...
-                                }
-                            }
+            fn from_owned_redis_value(v: Value) -> RedisResult<Self> {
+                FromRedisValue::from_owned_redis_value(v).map($func)
+            }
+        }
+    }
+}
+
+forwarded_impl!((T), Box<T>, Box::new);
```
Comment (nihohit) [inline redis/src/types.rs:1976]: why is it always `(T)` instead of `T`?

File: redis/src/types.rs [hunk 8]
```diff
@@ -1460,17 +1460,46 @@ impl<T: ToRedisArgs> ToRedisArgs for Option<T> {
...
 }
 
-impl<T: ToRedisArgs> ToRedisArgs for &T {
-    fn write_redis_args<W>(&self, out: &mut W)
-    where
-        W: ?Sized + RedisWrite,
-    {
-        (*self).write_redis_args(out)
-    }
+macro_rules! deref_impl {
```
Comment (nihohit) [inline redis/src/types.rs:1463]: deref_to_write_redis_args_impl

File: redis/src/types.rs [hunk 9]
```diff
@@ -1925,81 +1954,92 @@ impl FromRedisValue for String {
...
+            _ => invalid_type_error!(v, "Response type not vector compatible."),
         }
-    };
+    }
 }
 
-from_vec_from_redis_value!(<T> Vec<T>);
-from_vec_from_redis_value!(<T> std::sync::Arc<[T]>);
-from_vec_from_redis_value!(<T> Box<[T]>; Vec::into_boxed_slice);
+forwarded_impl!((T), Box<[T]>, Vec::into_boxed_slice);
```
Comment (nihohit) [inline redis/src/types.rs:2039]: move these to the other `forwarded_impl` declarations

File: redis/src/types.rs [hunk 10]
```diff
@@ -1925,81 +1954,92 @@ impl FromRedisValue for String {
...
 
-/// Implement `FromRedisValue` for `$Type` (which should use the generic parameter `$T`).
-///
-/// The implementation parses the value into a vec, and then passes the value through `$convert`.
-/// If `$convert` is ommited, it defaults to `Into::into`.
-macro_rules! from_vec_from_redis_value {
-    (<$T:ident> $Type:ty) => {
-        from_vec_from_redis_value!(<$T> $Type; Into::into);
-    };
+macro_rules! forwarded_impl {
```
Comment (nihohit) [inline redis/src/types.rs:1957]: pointer_from_redis_value_impl

**Review by nihohit — APPROVED** [review]
LGTM. @jaymell WDYT?

**Comment by xoac** [comment]
By removing `from_vec_from_redis_value!` it lose ability to know exact `type` when falling.. Currently I have no idea how to fix these

**Comment by nihohit** [comment]
hi @xoac,
if you're still stuck onb this, you can always split this PR to 2 - one with ToRedisArgs, the other with `FromRedisArgs`. Alternatively, you can just remove the problematic implementation (`Arc<T>`, right?) from this PR, and solve that particular problem later.

**Comment by xoac** [comment]
Yes, Just wanted to wait a little bit longer to check if maybe someone has better idea how to resolve issue (losing ability to return hint for type in error) that was added with [`f91f8e0` (#1088)](https://github.com/redis-rs/redis-rs/pull/1088/commits/f91f8e0e7fefbd4e070a9fccf48f16c385d9c2bd) The only solution from my side was to revert that commit what I did. 

Do you want to add more std types to be supported with above macros? Like `Mutx`, `Rc`, `RwMutex` etc?

**Comment by nihohit** [comment]
please add Rc. I'm not sure whether there's a usecase for mutex types, so I'll wait for a request for that.

File: redis/src/types.rs [hunk 11]
```diff
@@ -1975,6 +1979,7 @@ macro_rules! pointer_from_redis_value_impl {
 
 pointer_from_redis_value_impl!(T, Box<T>, Box::new);
 pointer_from_redis_value_impl!(T, std::sync::Arc<T>, std::sync::Arc::new);
+pointer_from_redis_value_impl!(T, std::rc::Rc<T>, std::rc::Rc::new);
```
[the remark this answers is not in the document]
Comment (xoac) [reply redis/src/types.rs:1982]: added

**Comment by xoac** [comment]
Anything left here from my side here to do?

**Comment by jaymell** [comment]
This looks great! I am a bit lost by this comment, though:
> I think this PR might contains breaking changes for some users because from now you need to specify output type. See test_vec function.

**Review by jaymell — APPROVED** [review]

**Comment by nihohit** [comment]
@jaymell see the changes in `test_arc_slice` (lines 204, 220, 222) - `Arc<[_]>` now needs to be explicitly typed, in order to disambiguate between it and `Arc<T>`. Not sure why this doesn't apply to `Box`.
